src/include/txn.i

```diff
@@ -262,36 +259,32 @@ __wt_txn_update_set_timestamp(WT_SESSION_IMPL *session,
 	wt_timestamp_t *timestamp;
 	bool needs_timestamp;
 
-	txn = &session->txn;
-	upd = op->u.op_upd;
 	if (was_set != NULL)
 		*was_set = false;
 
-	/*
-	 * The timestamp is in the page deleted structure for truncates, or
-	 * in the update for other operations.
-	 */
-	if (op->type == WT_TXN_OP_REF_DELETE)
-		timestamp = op->u.ref == NULL || op->u.ref->page_del == NULL ?
-		    NULL : &op->u.ref->page_del->timestamp;
-	else
-		timestamp = op->u.op_upd == NULL ?
-		    NULL : &op->u.op_upd->timestamp;
+	txn = &session->txn;
 
 	/*
 	 * Updates in the metadata never get timestamps (either now or at
 	 * commit): metadata cannot be read at a point in time, only the most
 	 * recently committed data matches files on disk.
 	 */
-	needs_timestamp = !WT_IS_METADATA(op->btree->dhandle) &&
-	    F_ISSET(txn, WT_TXN_HAS_TS_COMMIT) &&
-	    (timestamp == NULL || __wt_timestamp_iszero(timestamp) ||
-	    F_ISSET(txn, WT_TXN_PREPARE));
+	if (WT_IS_METADATA(op->btree->dhandle) ||
+	    !F_ISSET(txn, WT_TXN_HAS_TS_COMMIT))
+		return;
 
+	/*
+	 * The timestamp is in the page deleted structure for truncates, or
+	 * in the update for other operations.
+	 */
+	timestamp = op->type == WT_TXN_OP_REF_DELETE ?
+	    &op->u.ref->page_del->timestamp : &op->u.op_upd->timestamp;
+	needs_timestamp =
+	    __wt_timestamp_iszero(timestamp) || F_ISSET(txn, WT_TXN_PREPARE);
 	if (!needs_timestamp)
 		return;
 
-	if (for_prepare) {
+	if (for_prepare && op->type != WT_TXN_OP_REF_DELETE) {
```

Comment:
Nice catch - this makes me wonder if we should pull the code from `__wt_txn_commit` dealing with prepared `WT_TXN_OP_REF_DELETE` into here as well. I didn't because it also locks the ref, and I didn't think this was an obvious place to go looking for that.
@bvpvamsikrishna do you have an opinion about whether we should write a comment here describing why REF_DELETE isn't handled in this function, or moving the functionality in here?

Reply:
@agorrod, my opinion is to move functionality in here and rename the function as "__wt_txn_op_set_timestamp" (changed the update --> op).